compaction/compaction_manager.cc

```diff
@@ -1267,19 +1271,23 @@ class regular_compaction_task_executor : public compaction_task_executor, public
 
 }
 
-void compaction_manager::submit(table_state& t) {
+void compaction_manager::submit(table_state& t, std::optional<std::vector<sstables::shared_sstable>> candidates_opt) {
     if (t.is_auto_compaction_disabled_by_user()) {
```

Comment:
I remember we opted instead for decoupling the execution loop from task. Might be a problem with ICS which will work with runs only.

But do we really need it after we're no longer keeping tasks alive for 10s?

Usually, we want to retrieve candidate list close to execution as there might be changes to sstable set, or regular compaction itself can wait for major, and sstable set is very different thereafter. Also, we cannot hold candidate list without marking them as being compacted. Otherwise, another compaction attempt might pick then, and when we get to compact this candidate list, they are already deleted by former compaction.

Reply:
> But do we really need it after we're no longer keeping tasks alive for 10s?

Why would we want to create futile tasks if we know in advance there's nothing to do?

> Usually, we want to retrieve candidate list close to execution as there might be changes to sstable set, or regular compaction itself can wait for major, and sstable set is very different thereafter. Also, we cannot hold candidate list without marking them as being compacted. Otherwise, another compaction attempt might pick then, and when we get to compact this candidate list, they are already deleted by former compaction.

That's a valid point.  We can drop candidates_opt here in case it's not empty, so we only prevent empty tasks.